File: examples/Go/mdwrkapi.go
//  mdwrkapi class - Majordomo Protocol Worker API
//  Implements the MDP/Worker spec at http://rfc.zeromq.org/spec:7.
//
//  Author: iano <scaly.iano@gmail.com>
//  Based on C & Python example

package main

import (
	zmq "github.com/alecthomas/gozmq"
	"log"
	"time"
)

type Worker interface {
	Close()
	Recv([][]byte) [][]byte
}

type mdWorker struct {
	broker  string
	context *zmq.Context
	service string
	verbose bool
	worker  *zmq.Socket

	heartbeat   time.Duration
	heartbeatAt time.Time
	liveness    int
	reconnect   time.Duration

	expectReply bool
	replyTo     []byte
}

func NewWorker(broker, service string, verbose bool) Worker {
	context, _ := zmq.NewContext()
	self := &mdWorker{
		broker:    broker,
		context:   context,
		service:   service,
		verbose:   verbose,
		heartbeat: 2500 * time.Millisecond,
		liveness:  0,
		reconnect: 2500 * time.Millisecond,
	}
	self.reconnectToBroker()
	return self
}

func (self *mdWorker) reconnectToBroker() {
	if self.worker != nil {
		self.worker.Close()
	}
	self.worker, _ = self.context.NewSocket(zmq.DEALER)
	self.worker.SetLinger(0)
	self.worker.Connect(self.broker)
	if self.verbose {
		log.Printf("I: connecting to broker at %s...\n", self.broker)
	}
	self.sendToBroker(MDPW_READY, []byte(self.service), nil)
	self.liveness = HEARTBEAT_LIVENESS
	self.heartbeatAt = time.Now().Add(self.heartbeat)
}

func (self *mdWorker) sendToBroker(command string, option []byte, msg [][]byte) {
	if len(option) > 0 {
		msg = append([][]byte{option}, msg...)
	}

	msg = append([][]byte{nil, []byte(MDPW_WORKER), []byte(command)}, msg...)
	if self.verbose {
		log.Printf("I: sending %X to broker\n", command)
		Dump(msg)
	}
	self.worker.SendMultipart(msg, 0)
}

func (self *mdWorker) Close() {
	if self.worker != nil {
		self.worker.Close()
	}
	self.context.Close()
}

func (self *mdWorker) Recv(reply [][]byte) (msg [][]byte) {
	//  Format and send the reply if we were provided one

	if len(reply) == 0 && self.expectReply {
		panic("Error reply")
	}

	if len(reply) > 0 {
		if len(self.replyTo) == 0 {
			panic("Error replyTo")
		}
		reply = append([][]byte{self.replyTo, nil}, reply...)
		self.sendToBroker(MDPW_REPLY, nil, reply)
	}

	self.expectReply = true

	for {
		items := zmq.PollItems{
			zmq.PollItem{Socket: self.worker, Events: zmq.POLLIN},
		}

		_, err := zmq.Poll(items, self.heartbeat)
		if err != nil {
			panic(err) //  Interrupted
		}

		if item := items[0]; item.REvents&zmq.POLLIN != 0 {
			msg, _ = self.worker.RecvMultipart(0)
			if self.verbose {
				log.Println("I: received message from broker: ")
				Dump(msg)
			}
			self.liveness = HEARTBEAT_LIVENESS
			if len(msg) < 3 {
				panic("Invalid msg") //  Interrupted
			}

			header := msg[1]
			if string(header) != MDPW_WORKER {
				panic("Invalid header") //  Interrupted
			}

			switch command := string(msg[2]); command {
			case MDPW_REQUEST:
				//  We should pop and save as many addresses as there are
				//  up to a null part, but for now, just save one...
				self.replyTo = msg[3]
				msg = msg[5:]
				return
			case MDPW_HEARTBEAT:
				// do nothin
			case MDPW_DISCONNECT:
				self.reconnectToBroker()
			default:
				log.Println("E: invalid input message:")
				Dump(msg)
			}
		} else if self.liveness--; self.liveness <= 0 {
			if self.verbose {
				log.Println("W: disconnected from broker - retrying...")
			}
			time.Sleep(self.reconnect)
			self.reconnectToBroker()
		}

		//  Send HEARTBEAT if it's time
		if self.heartbeatAt.Before(time.Now()) {
			self.sendToBroker(MDPW_HEARTBEAT, nil, nil)
			self.heartbeatAt = time.Now().Add(self.heartbeat)
		}
	}

	return
}

File: examples/C#/mdwrkapi.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading;

using ZeroMQ;

namespace Examples
{
	namespace MDWrkApi
	{
		//
		//  mdwrkapi class - Majordomo Protocol Worker API
		//  Implements the MDP/Worker spec at http://rfc.zeromq.org/spec:7.
		//
		// Author: chubbson
		//

		public class MajordomoWorker : IDisposable
		{
			//  Structure of our class
			//  We access these properties only via class methods

			// Our context
			private ZContext _context;

			// Majordomo broker
			public string Broker { get; protected set; }
			public string Service { get; protected set; }

			//  Socket to broker
			public ZSocket Worker { get; protected set; }

			//  Print activity to console
			public bool Verbose { get; protected set; }


			#region Heartbeat management

			// When to send HEARTBEAT
			public DateTime HeartbeatAt { get; protected set; }

			// How many attempts left
			public UInt64 Liveness { get; protected set; }

			// Heartbeat delay, msecs
			public TimeSpan Heartbeat { get; protected set; }

			// Reconnect delay, msecs
			public TimeSpan Reconnect { get; protected set; }

			#endregion

			#region Unknown to port

			// Zero only at start
			private bool _expectReply;

			// Return identity, if any
			private ZFrame _replyTo;

			#endregion


			public void SendToBroker(string command, string option, ZMessage msg)
			{
				using (msg = msg != null ? msg.Duplicate() : new ZMessage())
				{
					if (!string.IsNullOrEmpty(option))
						msg.Prepend(new ZFrame(option));
					msg.Prepend(new ZFrame(command));
					msg.Prepend(new ZFrame(MdpCommon.MDPW_WORKER));
					msg.Prepend(new ZFrame(string.Empty));

					if (Verbose)
						msg.DumpZmsg("I: sending '{0:X}|{0}' to broker", command.ToMdCmd());

					Worker.Send(msg);
				}
			}

			public void ConnectToBroker()
			{
				//  Connect or reconnect to broker
				Worker = new ZSocket(_context, ZSocketType.DEALER);
				Worker.Connect(Broker);
				if (Verbose)
					"I: connecting to broker at {0}...".DumpString(Broker);

				// Register service with broker
				SendToBroker(MdpCommon.MdpwCmd.READY.ToHexString(), Service, null);

				// if liveness hits zero, queue is considered disconnected
				Liveness = MdpCommon.HEARTBEAT_LIVENESS; 
				HeartbeatAt = DateTime.UtcNow + Heartbeat;
			}

			public MajordomoWorker(string broker, string service, bool verbose)
			{
				if(broker == null)
					throw new InvalidOperationException();
				if(service == null)
					throw new InvalidOperationException();

				_context = new ZContext();
				Broker = broker;
				Service = service;
				Verbose = verbose;
				Heartbeat = MdpCommon.HEARTBEAT_DELAY;
				Reconnect = MdpCommon.RECONNECT_DELAY;

				ConnectToBroker();
			}

			~MajordomoWorker()
			{
				Dispose(false);
			}

			public void Dispose()
			{
				GC.SuppressFinalize(this);
				Dispose(true);
			}

			protected void Dispose(bool disposing)
			{
				if (disposing)
				{
					// Destructor

					if (Worker != null)
					{
						Worker.Dispose();
						Worker = null;
					}
					//Do not Dispose Context: cuz of weird shutdown behavior, stucks in using calls
				}
			}

			//  .split configure worker
			//  We provide two methods to configure the worker API. You can set the
			//  heartbeat interval and retries to match the expected network performance.

			//  Set heartbeat delay
			public void Set_Heartbeat(int heartbeatInMs)
			{
				Heartbeat = TimeSpan.FromMilliseconds(heartbeatInMs);
			}

			//  Set reconnect delay
			public void Set_Reconnect(int reconnectInMs)
			{
				Reconnect = TimeSpan.FromMilliseconds(reconnectInMs);
			}


			//  .split recv method
			//  This is the {{recv}} method; it's a little misnamed because it first sends
			//  any reply and then waits for a new request. If you have a better name
			//  for this, let me know.

			//  Send reply, if any, to broker and wait for next request.
			public ZMessage Recv(ZMessage reply, CancellationTokenSource cancellor)
			{
				if (reply == null
					&& _expectReply)
					throw new InvalidOperationException();

				if (reply != null)
				{
					if(_replyTo == null)
						throw new InvalidOperationException();
					reply.Wrap(_replyTo);
					SendToBroker(MdpCommon.MdpwCmd.REPLY.ToHexString(), string.Empty, reply);
				}
				_expectReply = true;

				while (true)
				{
					if (cancellor.IsCancellationRequested
						|| (Console.KeyAvailable && Console.ReadKey(true).Key == ConsoleKey.Escape))
						_context.Shutdown();

					var p = ZPollItem.CreateReceiver();
					ZMessage msg;
					ZError error;
					if (Worker.PollIn(p, out msg, out error, Heartbeat))
					{
						using (msg)
						{
							// If we got a reply
							if (Verbose)
								msg.DumpZmsg("I: received message from broker:");

							Liveness = MdpCommon.HEARTBEAT_LIVENESS;

							// Don't try to handle errors, just assert noisily
							if(msg.Count < 3)
								throw new InvalidOperationException();

							using (ZFrame empty = msg.Pop())
							{
								if (!empty.ToString().Equals(""))
									throw new InvalidOperationException();
							}

							using (ZFrame header = msg.Pop())
							{
								if (!header.ToString().Equals(MdpCommon.MDPW_WORKER))
									throw new InvalidOperationException();
							}
							//header.ReadString().Equals(MDPW_WORKER);
							using (ZFrame command = msg.Pop())
							{
								if (command.StrHexEq(MdpCommon.MdpwCmd.REQUEST))
								{
									//  We should pop and save as many addresses as there are
									//  up to a null part, but for now, just save one...
									_replyTo = msg.Unwrap();

									//  .split process message
									//  Here is where we actually have a message to process; we
									//  return it to the caller application:
									return msg.Duplicate();
								}
								else if (command.StrHexEq(MdpCommon.MdpwCmd.HEARTBEAT))
								{
									// Do nothing for heartbeats
								}
								else if (command.StrHexEq(MdpCommon.MdpwCmd.DISCONNECT))
								{
									ConnectToBroker();
								}
								else
									"E: invalid input message: '{0}'".DumpString(command.ToString());
							}
						}
					}
					else if (Equals(error, ZError.ETERM))
					{
						cancellor.Cancel();
						break; // Interrupted
					}
					else if (Equals(error, ZError.EAGAIN) 
							 && --Liveness == 0)
					{
						if (Verbose)
							"W: disconnected from broker - retrying...".DumpString();
						Thread.Sleep(Reconnect);
						ConnectToBroker();
					}

					// Send HEARTBEAT if it's time
					if (DateTime.UtcNow > HeartbeatAt)
					{
						SendToBroker(MdpCommon.MdpwCmd.HEARTBEAT.ToHexString(), null, null);
						HeartbeatAt = DateTime.UtcNow + Heartbeat;
					}
				}
				if (cancellor.IsCancellationRequested)
					"W: interrupt received, killing worker...\n".DumpString();

				return null;
			}
		}
	}
}
